c compiler bug: V compiler doesn't check if a c struct can be casted to int

### Describe the bug

Hi, so i was using raylib, and i goofed up and tried to cast a raylib strict to an int. But the v compiler didnt gave me any errors. 


### Reproduction Steps

Basically what you need to do to reproduce it is a v bindings to a c struct, and then cast the external struct to  int() or f32 (or other  number). Casting to a string  errors correctly.

Rapid test:
```v install raylib```
past this into a v file
```v

module main

import raylib

struct Ball {
mut:
	position raylib.Vector3
	hp       int
}

fn main() {
	ow := Ball{
		position: raylib.Vector3{
			x: 0.0
			y: 1.7
			z: 0.0
		}
		hp:       0
	}
	// Here is the error,
	println(int(ow.position))
}

```

and  run
 ```v run main.v```

### Expected Behavior

A v compiler error, saying that a struct cannot be casted to an int
```
src/main.v:21:10: error: cannot cast struct `Vector3` to `int`
   19 |     }
   20 |     // Here is the error,
   21 |     println(int(ow.position))
      |             ~~~~~~~
   22 | }
```
### Current Behavior
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/bug.01JMA08EMVT60G1XPVS2WM0D7R.tmp.c:6126: error: cannot convert 'struct Vector3' to 'int'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
### Possible Solution
no clue (sry)


### Additional Information/Context

I was using v raylib bindings, but i reproduced it with sdl bindings. This is a bug with any v bindings that have c struct

### V version

V 0.4.9 4c8c892.2e421ba

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.9 4c8c892.2e421ba
|:-------------------|:-------------------
|OS                  |linux, "CachyOS"
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 Pro 7735U with Radeon Graphics
|Memory              |3.45GB/30.12GB
|                    |
|V executable        |/home/polyface/git/v/v
|V last modified time|2025-02-17 12:06:52
|                    |
|V home dir          |OK, value: /home/polyface/git/v
|VMODULES            |OK, value: /home/polyface/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/polyface/Documents/v/bug
|                    |
|Git version         |git version 2.48.1
|V git status        |2e421bac-dirty
|.git/config present |true
|                    |
|cc version          |cc (GCC) 14.2.1 20250207
|gcc version         |gcc (GCC) 14.2.1 20250207
|clang version       |clang version 19.1.7
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.41

The problem is not related to the library mentioned. I didn't know, but now I assume in general any alias of a C struct when tried to cast incorrectly to another type would give a C error. Next reduction code gives a C error because `Vector3` is not a pure V object. Thoughts?
```v
module main

#flag -I .
#include "vectors.h"

// content of vectors.h
// struct Vector3 {
// 	float x;
// 	float y;
//	float z;
// };

pub type Vector3 = C.Vector3
pub struct C.Vector3 {
	x f32
	y f32
	z f32
}

fn main() {
	v := Vector3 {
		x: 0.0
		y: 1.7
		z: 0.0
	}
	dump(v)
	dump(int(v)) // C error
}
```
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/23748.01JMB04G9W532XWQ9NYYVCTJWM.tmp.c:6007: error: cannot convert 'struct Vector3' to 'int'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
```

The problem here is because of the type alias of the C type, especially if the alias comes before the V definition.
```v
module main

#flag -I .
#include "vectors.h"

pub struct C.Vector3 {
        x f32
        y f32
        z f32
}

fn main() {
        v := C.Vector3 {
                x: 0.0
                y: 1.7
                z: 0.0
        }
        dump(v)
        dump(int(v)) // C error
}
```
gives the correct error when you try to compile:
```
x.v:19:7: error: cannot cast struct `C.Vector3` to `int`
   17 |     }
   18 |     dump(v)
   19 |     dump(int(v)) // C error
      |          ~~~~~~
   20 | }
```
If you move the type alias _under_ the V definition, you get a whole different error:
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/x.01JMB5RT7H0EQFHVPXE2E2C901.tmp.c:1711: error: dereferencing incomplete type 'Vector3'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
...
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3568,9 +3568,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
-	}
-
-	if to_sym.kind == .rune && from_sym.is_string() {
+	} else if to_sym.kind == .rune && from_sym.is_string() {
 		snexpr := node.expr.str()
 		ft := c.table.type_to_str(from_type)
 		c.error('cannot cast `${ft}` to rune, use `${snexpr}.runes()` instead.', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3568,7 +3568,8 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
-	} else if final_from_sym.kind == .struct && to_sym.kind !in [final_from_sym.kind, .voidptr] {
+	} else if !from_type.is_ptr() && final_from_sym.kind == .struct
+		&& to_sym.kind !in [.struct, .voidptr] {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3569,7 +3569,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
 	} else if !from_type.is_ptr() && final_from_sym.kind == .struct
-		&& to_sym.kind !in [.struct, .voidptr] {
+		&& final_to_sym.kind !in [.struct, .voidptr] {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3568,6 +3568,10 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
+	} else if final_from_sym.kind == .struct && to_sym.kind !in [final_from_sym.kind, .voidptr] {
+		ft := c.table.type_to_str(from_type)
+		tt := c.table.type_to_str(to_type)
+		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)
 	}
 
 	if to_sym.kind == .rune && from_sym.is_string() {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3568,17 +3568,17 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
-	} else if !from_type.is_ptr() && final_from_sym.info is ast.Struct
-		&& !final_from_sym.info.is_empty_struct() && final_to_sym.is_int() {
-		ft := c.table.type_to_str(from_type)
-		tt := c.table.type_to_str(to_type)
-		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)
 	}
 
 	if to_sym.kind == .rune && from_sym.is_string() {
 		snexpr := node.expr.str()
 		ft := c.table.type_to_str(from_type)
 		c.error('cannot cast `${ft}` to rune, use `${snexpr}.runes()` instead.', node.pos)
+	} else if !from_type.is_ptr() && final_from_sym.info is ast.Struct
+		&& !final_from_sym.info.is_empty_struct() && final_to_sym.is_int() {
+		ft := c.table.type_to_str(from_type)
+		tt := c.table.type_to_str(to_type)
+		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)
 	}
 
 	if to_sym.kind == .enum && !(c.inside_unsafe || c.file.is_translated) && from_sym.is_int() {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3569,8 +3569,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
 	} else if !from_type.is_ptr() && final_from_sym.info is ast.Struct
-		&& !final_from_sym.info.is_empty_struct() && final_to_sym.kind != .rune
-		&& final_to_sym.is_int() {
+		&& !final_from_sym.info.is_empty_struct() && final_to_sym.is_int() {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3568,7 +3568,9 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
-	} else if !from_type.is_ptr() && final_from_sym.kind == .struct && final_to_sym.is_int() {
+	} else if !from_type.is_ptr() && final_from_sym.info is ast.Struct
+		&& !final_from_sym.info.is_empty_struct() && final_to_sym.kind != .rune
+		&& final_to_sym.is_int() {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3568,8 +3568,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast incompatible option ${final_to_sym.name} `${ft}` to `${tt}`',
 			node.pos)
-	} else if !from_type.is_ptr() && final_from_sym.kind == .struct
-		&& final_to_sym.kind !in [.struct, .voidptr] {
+	} else if !from_type.is_ptr() && final_from_sym.kind == .struct && final_to_sym.is_int() {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3574,8 +3574,9 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		snexpr := node.expr.str()
 		ft := c.table.type_to_str(from_type)
 		c.error('cannot cast `${ft}` to rune, use `${snexpr}.runes()` instead.', node.pos)
-	} else if !from_type.is_ptr() && final_from_sym.info is ast.Struct
-		&& !final_from_sym.info.is_empty_struct() && final_to_sym.is_int() {
+	} else if !from_type.is_ptr() && from_type != ast.string_type
+		&& final_from_sym.info is ast.Struct && !final_from_sym.info.is_empty_struct()
+		&& final_to_sym.is_int() {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3574,7 +3574,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast `${ft}` to rune, use `${snexpr}.runes()` instead.', node.pos)
 	} else if !from_type.is_ptr() && from_type != ast.string_type
 		&& final_from_sym.info is ast.Struct && !final_from_sym.info.is_empty_struct()
-		&& final_to_sym.is_int() {
+		&& (final_to_sym.is_int() || final_to_sym.is_float()) {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast type `${ft}` to `${tt}`', node.pos)

--- a/vlib/v/checker/tests/c_struct_cast.out
+++ b/vlib/v/checker/tests/c_struct_cast.out
@@ -1,0 +1,6 @@
+vlib/v/checker/tests/c_struct_cast.vv:18:7: error: cannot cast type `Vector3` to `int`
+   16 |     }
+   17 |     dump(v)
+   18 |     dump(int(v))
+      |          ~~~~~~
+   19 | }

--- a/vlib/v/checker/tests/c_struct_cast.vv
+++ b/vlib/v/checker/tests/c_struct_cast.vv
@@ -1,0 +1,19 @@
+module main
+
+pub struct C.Vector3 {
+	x f32
+	y f32
+	z f32
+}
+
+pub type Vector3 = C.Vector3
+
+fn main() {
+	v := Vector3{
+		x: 0.0
+		y: 1.7
+		z: 0.0
+	}
+	dump(v)
+	dump(int(v))
+}

